doc/developer-guides/sw_design_guidelines.rst

```diff
@@ -0,0 +1,494 @@
+   |                    +-------------------------+--------------+---------------------------+-------------------------+
+   |                    | Invalid hypercall       | Yes          | The hypervisor shall      | Invalid hypercall       |
+   |                    | parameter               |              | return related error code | parameter provided by   |
+   |                    |                         |              | to the VM.                | any VM.                 |
+   |                    +-------------------------+--------------+---------------------------+-------------------------+
+   |                    | Invalid data in the     | Yes          | Case by case depending    | Invalid data in memory  |
+   |                    | sharing memory area     |              | on the data.              | shared with all VMs,    |
+   |                    |                         |              |                           | such as IO request      |
+   |                    |                         |              |                           | buffers and sbuf for    |
+   |                    |                         |              |                           | debug.                  |
+   +--------------------+-------------------------+--------------+---------------------------+-------------------------+
+   | External resource  | Invalid E820 table or   | Yes          | The hypervisor shall      | Invalid E820 table or   |
+   | provided by        | invalid boot information|              | panic during platform     | invalid boot information|
+   | bootloader         |                         |              | initialization.           | .                       |
+   | (UEFI or SBL)      |                         |              |                           |                         |
+   +--------------------+-------------------------+--------------+---------------------------+-------------------------+
+   | Physical resource  | 1GB page is not         | Yes          | The hypervisor shall      | 1GB page is not         |
+   | used by the        | available on the        |              | panic during platform     | available on the        |
+   | hypervisor         | platform or invalid     |              | initialization.           | platform or invalid     |
+   |                    | physical CPU ID         |              |                           | physical CPU ID.        |
+   +--------------------+-------------------------+--------------+---------------------------+-------------------------+
+
+
+Examples
+--------
+
+Here is an example to illustrate when error handling codes are required on
+architecture level.
+
+There are two pre-condition statements of ``vcpu_from_vid``. It indicates that
+it's the caller's responsibility to guarantee these pre-conditions.
+
+.. code-block:: c
+
+  /*
+   * @pre vcpu_id < CONFIG_MAX_VCPUS_PER_VM
```

Comment:
It looks like you're using the ``@pre`` doxygen comment here, but the comment isn't properly written as a doxygen comment (with a ``/**`` beginning): 
```
/**
 * @pre vcpu_id < CONFIG_MAX_VCPUS_PER_VM
 * @pre &(vm->hw.vcpu_array[vcpu_id])->state != VCPU_OFFLINE
 */
```